feature(builtins): add a new builtin function to render templated strings

RDVasavada opened the pull request:
<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see our contributor guide below.

For more information on contributing to OPA see:

* [Contributing Guide](https://www.openpolicyagent.org/docs/latest/contributing/)
  for high-level contributing guidelines and development setup.

-->

### Why the changes in this PR are needed?
closes: #6371 

Adds a builtin function for template rendering.
<!--
Include a short description of WHY the changes were made.
-->

### What are the changes in this PR?
Adds a new builtin function that accepts a templated string and a mapping of vars to inject, and returns the rendered string with the vars injected.

<!--
Include a short description of WHAT changes were made.
-->

### Notes to assist PR review:
For this builtin function, I've set the `missingkey` option to `error`, so that if the template contains a variable that is not present in the given variable mapping, the function will not attempt to render the template and return an error. This was done to prevent silent failures from occurring. 
  

<!--
Here you can add information you think will help the reviewer(s).
-->


<!--
Here you can include links to additional resources related to the changes, discuss your solution, other approaches you considered etc.
-->

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -1715,6 +1716,19 @@ var JSONIsValid = &Builtin{
 
+var RenderTemplate = &Builtin{
+	Name:        "render_template",
+	Description: "Renders a templated string with given template variables injected. See https://pkg.go.dev/text/template for templating examples.",
```
Nit: If you can explain a bit what the built-in does instead of the including a link that would be better.

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -1715,6 +1716,19 @@ var JSONIsValid = &Builtin{
 }
 
+var RenderTemplate = &Builtin{
+	Name:        "render_template",
```
This is currently under the `encoding` section. Is this better suited under `strings`. For ex `strings.render_template`

RDVasavada replied:
Agreed, I think it does make more sense under strings

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -1715,6 +1716,19 @@ var JSONIsValid = &Builtin{
+	Description: "Renders a templated string with given template variables injected. See https://pkg.go.dev/text/template for templating examples.",
+	Decl: types.NewFunction(
+		types.Args(
+			types.Named("x", types.S).Description("a string"),
```
Nit: Can we elaborate the desc a bit. For example `a templated string` or something.

ashutosh-narkar commented on test/cases/testdata/rendertemplate/rendertemplate.yaml:
```diff
@@ -0,0 +1,24 @@
+  query: data.test.p = x
+  modules:
+  - |
+    package test
```
It would be helpful to add an error case. eg. missing keys.

ashutosh-narkar reviewed (commented):
@RDVasavada thanks for the contribution. Few comments inline.

RDVasavada commented:
Thanks for your feedback @ashutosh-narkar. I've addressed your comments -- let me know if you have any additional feedback!

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -1317,6 +1318,22 @@ var StringReverse = &Builtin{
+	`,
+	Decl: types.NewFunction(
+		types.Args(
+			types.Named("str", types.S).Description("a templated string"),
```
For consistency, replace `str` -> `value`

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -1317,6 +1318,22 @@ var StringReverse = &Builtin{
+	Decl: types.NewFunction(
+		types.Args(
+			types.Named("str", types.S).Description("a templated string"),
+			types.Named("vars", types.A).Description("a mapping of template variable keys to values"),
```
Shouldn't this be an object (`types.NewObject`)?

RDVasavada replied:
Yes, thanks for the suggestion

ashutosh-narkar commented on ast/builtins.go:
```diff
@@ -1317,6 +1318,22 @@ var StringReverse = &Builtin{
+	Description: `Renders a templated string with given template variables injected. For a given template string and key/value mapping, values will be injected into the template where they are referenced by key. 
+	The templated string can contain simple syntax (such as "Hello {{.name}}") or complex syntax (such as "{{if .shouldsayhello}}Hello world!{{end}}").
+	For more examples, see https://pkg.go.dev/text/template 
+	`,
```
We can probably remove this `The templated string can contain simple syntax (such as "Hello {{.name}}") or complex syntax (such as "{{if .shouldsayhello}}Hello world!{{end}}").
	For more examples, see https://pkg.go.dev/text/template`

RDVasavada commented on ast/builtins.go:
```diff
@@ -1317,6 +1318,22 @@ var StringReverse = &Builtin{
+		),
+		types.Named("result", types.A).Description("rendered template with template variables injected"),
+	),
+	Categories: encoding,
```
Good catch, thanks

ashutosh-narkar commented on test/cases/testdata/rendertemplate/rendertemplate.yaml:
```diff
@@ -0,0 +1,37 @@
+    package test
+
+    template_string = `{{.test}}`
+    template_vars = {`test`: `hello world`}
```
@RDVasavada the value does not have to be string, correct?

RDVasavada replied:
That's correct

ashutosh-narkar reviewed (approved):
LGTM

ashutosh-narkar commented:
@RDVasavada can you please squash your commits and we can get this in. Thanks!

ashutosh-narkar commented:
Check [this](https://www.openpolicyagent.org/docs/latest/contrib-code/#commit-messages) link if you need pointers to craft a good commit message.

johanfylling commented on builtin_metadata.json:
```diff
@@ -15898,6 +15899,31 @@
+    "available": [
+      "edge"
+    ],
+    "description": "Renders a templated string with given template variables injected. For a given template string and key/value mapping, values will be injected into the template where they are referenced by key.",
```
Maybe we should link to the go text/template [docs](https://pkg.go.dev/text/template) here, as the word `template` doesn't give many clues to what that actually entails.

johanfylling replied:
The change would need to go into `ast/builtins.go`, of course 😄 .

RDVasavada commented:
Added a unit test for non-string template var inputs and added back the template docs link as requested by @johanfylling, please let me know if there's any other feedback before I squash, thanks!

ashutosh-narkar reviewed (commented):
LGTM.

RDVasavada commented:
@ashutosh-narkar rebased and squashed
